integration/client/container_windows_test.go

```diff
@@ -0,0 +1,72 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package client
+
+import (
+	"golang.org/x/sys/windows"
+)
+
+// grantReadToEveryone adds a new entry in the ACL
+// of the file or folder, allowing "Everyone" to read
+// the contents. Inherit is set for all objects and
+// containers.
+func grantReadToEveryone(file string) error {
```

Comment:
I'm confused why you would need to do this manually. Does Docker do this and I didnt know for a passed in mount? This PR would imply that any folder share would not be read capable to ContainerUser by default right?

Reply:
Hi @jterry75 ,

This test is meant to validate that given an arbitrary folder, passed in as a volume to a container, the container can read or write inside of it, given the proper permissions.

Filesystem permissions are enforced by the host filesystem. 

Traditionally, to allow users inside containers to write to a volume on Linux, the user would have to make sure the folder being mounted from the host had the proper owner or the permissions allowed the user inside of the container to write to the folder being mounted. That means that before using the folder as a volume in the container, the user would first determine the UID of the user inside the container, as seen by the host, and then issue a ```chown``` on the folder. Unprivileged containers in Linux will map container UIDs to a different range of UIDs on the host. So the container would see UID 0, but on the host that would translate to (for example) ```1000000```. In such cases, the user would need to change the owner on the folder to UID ```1000000``` on the host, to allow the root user inside the container to write to the volume (or allow "others" write access - ie: chmod 777).

The same thing happens in Windows as well, but instead of UIDs and POSIX permissions, we have SIDs and ACLs. We could conceivably not grant write to everyone in this test, and just grant it to the well known SID of ```ContainerUser``` but I am not sure it's worth adding a condition which may break in the future, if the image being used will run as a different user.

That being said, this test simulates situations in which the user would create a folder, grant appropriate permissions and then spawn a container with that folder mounted as a volume. The user would then attempt to read and write to that volume.

Changing permissions automatically on a folder mounted by the user is not desirable (I think), as it may not be what the user wants, and would mean we create a security issue. Setting permissions does happen automatically for volumes declared in the ```Dockerfile``` via the ```VOLUME``` stanza (https://github.com/containerd/containerd/blob/main/integration/volume_copy_up_test.go#L96), but that is a different condition we are testing.

Reply:
@kevpar / @dcantah - This is not my area of expertise but I don't think this is correct for Windows. What @gabriel-samfira said above is 100% true about accounts with different names having different SIDs across machines (which is also true for containers). However, what doesn't seem true (I am guessing) is that we need to provide this capability in order to read/write to the folder. Permissions are specified at the time of the mount and it is the host that manages that not by some sort of filesystem SID matching to my knowledge. If it was, we would have to know the SID of CAdmin or CUser and write the ACL into the NTFS rights on the folder being shared before starting the container. I don't have the answer here but anecdotally based on Docker never needing this something is wrong here :).

Reply:
Note, there is a difference between permission specified when mounting an arbitrary folder as a volume inside a container, and automatically setting filesystem permissions on that folder.

Automatically allowing filesystem read/write when mounting, may allow one user of the system to read files from other users of the system

e.g.: User ```gabriel``` mounting ```c:\users\Administrator``` inside their own container, and gaining access to all private files.

Hence the need to properly set up permissions of the folder a user intends to use as a volume mount.

Any thoughts are welcome.

Reply:
One more note I forgot to mention a while back. ContainerAdministrator (S-1-5-93-2-1) and ContainerUser (S-1-5-93-2-2) have well known SIDs. These SIDs are the same on all installations of Windows.